[SymmetricMemory] allow overlapping devices for testing

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139677
* #139529
* #140129
* #140128
* __->__ #140127

When `TORCH_SYMM_MEM_ALLOW_OVERLAPPING_DEVICES` is set, the check for overlapping devices and multicast support will be disabled. This is useful for testing with a single device.

Making this is an env var instead of an API argument since this is likely only useful for testing.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
[SymmetricMemory] allow overlapping devices for testing

[ghstack-poisoned]

lw commented on test/distributed/test_symmetric_memory.py:
```diff
@@ -287,6 +287,41 @@ def test_wait_signal_timeout(self) -> None:
+    @skipIfRocm
+    @requires_cuda
+    def test_allow_overlapping_devices(self) -> None:
+        os.environ["TORCH_SYMM_MEM_ALLOW_OVERLAPPING_DEVICES"] = "1"
```
How do distributed tests work in PyTorch? Do we need to clear this env var at the end of the test case to avoid it "polluting" other test cases in case the same process gets reused?

yifuwang replied:
Thanks for checking! This test class is a `MultiProcessTestCase` which runs test methods in subprocesses.

lw reviewed (approved):
Thanks!

Commit:
Update

[ghstack-poisoned]

yifuwang commented:
@pytorchbot rebase

Commit:
Update

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] allow overlapping devices for testing"


When `TORCH_SYMM_MEM_ALLOW_OVERLAPPING_DEVICES` is set, the check for overlapping devices and multicast support will be disabled. This is useful for testing with a single device.

Making this is an env var instead of an API argument since this is likely only useful for testing.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
